### ISSUE TITLE
Check for API-key parameter first before requiring headers with remote authentication (15.07 with SSO)

### ISSUE THREAD

**Issue by MatthewRalston** [galaxyproject/galaxy#972]
Hello everyone,
I have a Galaxy server running behind a SSO/SiteMinder proxy for authentication. When our Apache server sees a request for galaxy.mydomain.com, it reroutes to our SSO for authentication. Once a cookie is present, apache reroutes to myserver.mydomain.com:galaxyport, passing the appropriate header to galaxy. Galaxy (to my understanding) always expects the header components when this remote authentication is enabled. However, since galaxy always checks for these headers, even when a valid api key is passed, the API doesn't seem to work when the remote/external authentication is activated and produces the same errors regarding HTTP_GX_SECRET (first) from #827.

I want to do the following:
- Be authenticated from my cookie/SSO when I access galaxy.mydomain.com, where Apache provides the right headers to Galaxy behind the scenes.
- Use bioblend and the Galaxy API to provide automation for my team by simply passing the appropriate API key via bioblend.

**Comment by dannon** [comment]
@MatthewRalston Coincidentally I'm looking at this right now from another report -- Galaxy shouldn't be bothering to check against REMOTE_USER when an API key is present (this may be a new regression)

**Comment by MatthewRalston** [comment]
Thanks @dannon! This will give me some time to look in to AWS with CentOS

**Comment by hexylena** [comment]
@dannon yep, this has been a problem for us such that I've just ignored the API completely. Sorry, I should've looked into this sooner.

I'd love to see everyone given an API key automatically and to have remote_user ignored completely on api routes. 

> Galaxy shouldn't be bothering to check against REMOTE_USER when an API key is present

this isn't really workable I don't think. Either the user is using via the UI and already has an SSO ticket, or they're accessing via the API and using an API key. 

Web users may not have an API key and if they do it isn't sent in API requests. Cli users accessing the API will never get through the SSO

**Comment by dannon** [comment]
@erasche The issue you're describing wrt: CLI users requires extra routing on the proxy side, and there's not a good way around it.  What I'm trying to track down is the Galaxy-side bug where, even with this SSO bypass, API key auth is failing.

**Comment by hexylena** [comment]
@dannon ah, okay. Sorry, I'll file my issue separately.

**Comment by dannon** [comment]
@MatthewRalston So I may need a little more help reproducing this one.

My nginx conf which emulates a remote user setup is something like the following.  Note the /galaxy/api bypass, which does not include the remote user header.

```
    upstream galaxy_app {
        server localhost:8080;
    }
    server {
        location /galaxy/api {
            proxy_pass   http://galaxy_app;
            proxy_set_header   X-Forwarded-Host $host;
            proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;
        }
        location /galaxy {
            proxy_pass   http://galaxy_app;
            proxy_set_header   X-Forwarded-Host $host;
            proxy_set_header   X-Forwarded-For  $proxy_add_x_forwarded_for;
            proxy_set_header   REMOTE_USER foo@foo.com;
        }
        ...
```

And this appears to work fine, testing w/ curl:

```
curl http://localhost:8000/galaxy/api/libraries/\?key\=*****************
```

**Comment by MatthewRalston** [comment]
Thanks @dannon, this is helpful. We're using apache locally for the time being but will switch to nginx once things are in the cloud. I'll check this out tomorrow.

**Comment by dannon** [comment]
@MatthewRalston Awesome.  In the meantime, I can try to set up an Apache proxy that does something similar (but works).

**Comment by MatthewRalston** [comment]
@dannon would you mind posting an anonymized version of the relevant section from your `galaxy.ini`?

**Comment by dannon** [comment]
@MatthewRalston https://gist.github.com/41699291b99de8927950

**Comment by dannon** [comment]
@MatthewRalston Turns out this is an issue with the way headers were being set when bypassing external auth via proxy (the /api bypass, as in my config above) under Apache.

#1003 should hopefully resolve this for you.

**Comment by MatthewRalston** [comment]
Okay. Any idea when is 15.10 slated to be released? I'd love to get this into production, manually if need be.

**Comment by dannon** [comment]
In the next week, I'd say.  Several people are out at Genome Informatics, which I think bumped us a bit off schedule.

**Comment by MatthewRalston** [comment]
Thank you!

**Comment by MatthewRalston** [comment]
Looking into this now.

**Comment by MatthewRalston** [comment]
Hi everyone and @golharam, I'm still seeing the issue regarding the HTTP_GX_SECRET from #827. I merged 15.10 into our codebase within the last week. Our SSO is setup for galaxy.example.com after which, it routes to the application at example.com:1234. The offending request using bioblend API: 

`http://example.com:1234/api/libraries?key=abcdefg56789`

Stack trace

``` python
URL: http://example.com:1234/api/libraries?key=abcdefg56789
File '.../galaxy-v15.05-production/eggs/WebError-0.8a-py2.7.egg/weberror/evalexception/middleware.py', line 364 in respond
  app_iter = self.application(environ, detect_start_response)
File '.../galaxy_env/lib/python2.7/site-packages/paste/recursive.py', line 84 in __call__
  return self.application(environ, start_response)
File '.../galaxy-v15.05-staging/lib/galaxy/web/framework/middleware/remoteuser.py', line 78 in __call__
  if not safe_str_cmp(environ.get('HTTP_GX_SECRET'), self.config_secret_header):
File '.../galaxy-v15.05-staging/lib/galaxy/util/__init__.py', line 1244 in safe_str_cmp
  if len(a) != len(b):
TypeError: object of type 'NoneType' has no len()
```

Contents of lib/galaxy/web/framework/webapp.py

``` python
            if galaxy_session:
                # An existing session, make sure correct association exists
                if galaxy_session.user is None:
                    # No user, associate
                    galaxy_session.user = self.get_or_create_remote_user( remote_user_email )
                    galaxy_session_requires_flush = True
                elif (not remote_user_email.startswith('(null)') and  # Apache does this, see remoteuser.py
                      (galaxy_session.user.email != remote_user_email) and
                      ((not self.app.config.allow_user_impersonation) or
                       (remote_user_email not in self.app.config.admin_users_list))):
                    # Session exists but is not associated with the correct
                    # remote user, and the currently set remote_user is not a
                    # potentially impersonating admin.
                    invalidate_existing_session = True
                    user_for_new_session = self.get_or_create_remote_user( remote_user_email )
                    log.warning( "User logged in as '%s' externally, but has a cookie as '%s' invalidating session",
                                 remote_user_email, galaxy_session.user.email )
```

@erasche here is the relevant section of my lib/galaxy/web/framework/middleware/remoteuser.py

``` python
        if self.config_secret_header is not None:
            if not safe_str_cmp(environ.get('HTTP_GX_SECRET'), self.config_secret_header):
                title = "Access to Galaxy is denied"
                message = """
                Galaxy is configured to authenticate users via an external
                method (such as HTTP authentication in Apache), but an
                incorrect shared secret key was provided by the
                upstream (proxy) server.</p>
                <p>Please contact your local Galaxy administrator.  The
                variable <code>remote_user_secret</code> and
                <code>GX_SECRET</code> header must be set before you may
                access Galaxy.
                """
                return self.error( start_response, title, message )
```

**Comment by dannon** [comment]
@MatthewRalston Got this one reproduced, working on a fix now.

**Comment by dannon** [comment]
@MatthewRalston Can you share your current apache configuration? (if it has changed, since above)

**Comment by MatthewRalston** [comment]
Hi @dannon I am working on getting it from my sysadmin.

**Comment by MatthewRalston** [comment]
```
ServerAdmin someuser@mycompany.com
ServerName galaxy.mycompany.com
ErrorLog logs/error_log-galaxy
CustomLog logs/access_log-galaxy common
DocumentRoot "/apps/sys/apache-2.2.26/htdocs"
RewriteEngine on
RewriteLog logs/rewrite_log-galaxy
RewriteLogLevel 9
ProxyPass / http://galaxy.mycompany.com:18080/
ProxyPassReverse / http://galaxy.mycompany.com:18080/
    RewriteCond %{IS_SUBREQ} ^false$
    RewriteCond %{LA-U:REMOTE_USER} (.+)
    RewriteRule . - [E=RU:%1]
    RequestHeader set REMOTE_USER "%{HTTP_MAIL}e" env=HTTP_MAIL
    RequestHeader set GX_SECRET "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
    RequestHeader set REMOTE_PROXYING_CONFIG "someone@mycompany.com"
    <Directory />
        Options All
        AllowOverride None
        Order deny,allow
        Allow from all
       PerlAccessHandler Apache2::AuthSiteMinder
       HeaderName header.html
       ReadmeName readme.html
       IndexIgnore header.*
       IndexIgnore readme.*

    </Directory>

```

**Comment by MatthewRalston** [comment]
As of today 2/22/16 I'm still getting the issue with 15.10. 
Everything is the same in the galaxy codebase. To reproduce the issue
1. configure Apache and SSO
2. Set up the GX_SECRET and external authentication in the galaxy.ini
3. Visit the ProxyPass page and see the stack trace.
 Thanks @dannon.

**Comment by dannon** [comment]
This should have been closed with #1801.
